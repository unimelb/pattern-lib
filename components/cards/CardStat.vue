<template>
  <FocusWrapper
    :color="color">
    <div :class="classes">
      <div class="card__inner">
        <h3 class="card-header__heading">
          {{ stat }}
        </h3>
        <p>{{ excerpt }}</p>
      </div>
    </div>
  </FocusWrapper>
</template>

<script>
import FocusWrapper from '../focus-wrapper/FocusWrapper.vue';

export default {
  name: 'CardStat',
  components: { FocusWrapper },
  props: {
    color: {
      type: String,
      default: '',
    },
    vertical: {
      type: Boolean,
      default: false,
    },
    excerpt: {
      type: String,
      default: 'Lorem ipsum dolor sit amet, consectetur.',
    },
    stat: {
      type: String,
      default: '123456',
      validator: (value) => {
        const maxLength = 6;

        return value.length && value.length <= maxLength;
      },
    },
    cols: {
      type: Number,
      default: 3,
      validator: (value) => {
        const minColumns = 2;
        const maxColumns = 3;

        return value >= minColumns && value <= maxColumns;
      },
    },
    inverted: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classes() {
      return {
        'card--focus-box-vert': this.vertical,
        'card--stat btn-owner card bg-inverted': this.inverted,
        'card--stat btn-owner card bg-white': !this.inverted,
      };
    },
  },
};
</script>
