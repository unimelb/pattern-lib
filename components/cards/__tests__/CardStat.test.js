import {
  shallow, mount,
} from 'vue-test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import CardStat from '../CardStat.vue';

expect.extend(toHaveNoViolations);

describe('CardStat', () => {
  it('should match snapshot', () => {
    const result = shallow(CardStat).element;
    expect(result).toMatchSnapshot();
  });

  it('should have default props and correct types vertical/excerpt/stat/cols/inverted', () => {
    const wrapper = shallow(CardStat);
    const {
      vertical,
      excerpt,
      stat,
      cols,
      inverted,
    } = wrapper.vm.$options.props;

    expect(vertical.type).toBe(Boolean);
    expect(wrapper.props().vertical).toBe(false);

    expect(stat.type).toBe(String);
    expect(wrapper.props().stat).toBe('123456');

    expect(excerpt.type).toBe(String);
    expect(wrapper.props().excerpt).toBe('Lorem ipsum dolor sit amet, consectetur.');

    expect(cols.type).toBe(Number);
    expect(wrapper.props().cols).toBe(3);

    expect(inverted.type).toBe(Boolean);
    expect(wrapper.props().inverted).toBe(false);
  });

  it('should render stat from prop with correct type', () => {
    const stat = '123456';
    const wrapper = shallow(CardStat, {
      propsData: {
        stat,
      },
    });
    expect(typeof wrapper.props().stat).toBe('string');
    expect(wrapper.props().stat).toBe(stat);
  });

  it('should render excerpt from prop with correct type', () => {
    const excerpt = 'Lorem ipsum dolor sit amet, consectetur.';
    const wrapper = shallow(CardStat, {
      propsData: {
        excerpt,
      },
    });

    expect(typeof wrapper.props().excerpt).toBe('string');
    expect(wrapper.props().excerpt).toBe(excerpt);
  });

  it('should accept cols prop with correct type', () => {
    const cols = 3;
    const wrapper = shallow(CardStat, {
      propsData: {
        cols,
      },
    });

    expect(typeof wrapper.props().cols).toBe('number');
    expect(wrapper.props().cols).toBe(cols);
  });

  it('should accept cols max = 3 and min = 2', () => {
    const wrapper = shallow(CardStat);
    const {
      cols,
    } = wrapper.vm.$options.props;

    expect(cols.validator && cols.validator(1)).toBeFalsy();
    expect(cols.validator && cols.validator(4)).toBeFalsy();
  });

  it('Component throws no accessibility violations', (done) => {
    const html = mount(CardStat).html();
    // pass anything that outputs html to axe
    return axe(html).then((response) => {
      expect(response).toHaveNoViolations();
      done();
    });
  });
});
