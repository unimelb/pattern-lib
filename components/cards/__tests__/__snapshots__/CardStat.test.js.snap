// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardStat should match snapshot 1`] = `<!---->`;
